frontend/catalyst/utils/jax_extras.py

```diff
@@ -262,48 +282,68 @@ def pad_jaxpr_constvars(i, jaxpr):
-    """Wraps the callable ``f`` into a WrappedFun JAX container. Calculate input abstract values
-    and output_tree promise. The promise must be called after the resulting wrapped function is
-    evaluated."""
+    """Wraps the callable ``f`` into a WrappedFun container accepting collapsed flatten arguments
+    and returning expanded flatten results. Calculate input abstract values and output_tree promise.
+    The promise must be called after the resulting wrapped function is evaluated."""
     flat_args, in_tree = tree_flatten((args, kwargs))
+    abstracted_axes = None
+    axes_specs = _flat_axes_specs(abstracted_axes, *args, **kwargs)
+    in_type = infer_lambda_input_type(axes_specs, flat_args)
+    in_avals, keep_inputs = unzip2(in_type)
     wf = wrap_init(f)
-    in_avals, keep_inputs = list(map(shaped_abstractify, flat_args)), [True] * len(flat_args)
-    in_type = tuple(zip(in_avals, keep_inputs))
     wff, out_tree_promise = flatten_fun(wf, in_tree)
     wffa = annotate(wff, in_type)
-    return wffa, in_avals, out_tree_promise
+    return wffa, in_avals, keep_inputs, out_tree_promise
+
+
+def get_aval2(x):
+    """An extended version of `jax.core.get_aval` which also accepts AbstractValues."""
+    # TODO: remove this patch when https://github.com/google/jax/pull/18579 is merged
+    if isinstance(x, AbstractValue):
+        return x
+    elif isinstance(x, Tracer):
+        return x.aval
+    else:
+        return concrete_aval(x)
+
 
+def _no_clean_up_dead_vars(_eqn, _env, _last_used):
+    """A stub to workaround the Jax ``KeyError 'a'`` bug during the lowering of Jaxpr programs to
+    MLIR with the dynamic API enabled."""
+    return None
 
-def jaxpr_to_mlir(func_name, jaxpr, shape):
-    """Lower a Python function into an MLIR module.
+
+def jaxpr_to_mlir(func_name, jaxpr):
+    """Lower a Jaxpr into an MLIR module.
 
     Args:
-        func: python function to be lowered
-        args: arguments to ``func``
-        kwargs: keyword arguments to ``func``
+        func_name(str): function name
+        jaxpr(Jaxpr): Jaxpr code to lower
 
     Returns:
         module: the MLIR module corresponding to ``func``
         context: the MLIR context corresponding
-        jaxpr: the jaxpr corresponding to ``func``
-        shape: the shape of the return values in ``PyTreeDef``
     """
 
-    nrep = jaxpr_replicas(jaxpr)
-    effects = jax_ordered_effects.filter_in(jaxpr.effects)
-    axis_context = ReplicaAxisContext(xla.AxisEnv(nrep, (), ()))
-    name_stack = new_name_stack(wrap_name("ok", "jit"))
-    module, context = custom_lower_jaxpr_to_module(
-        func_name="jit_" + func_name,
-        module_name=func_name,
-        jaxpr=jaxpr,
-        effects=effects,
-        platform="cpu",
-        axis_context=axis_context,
-        name_stack=name_stack,
-    )
+    with Patcher(
+        (jax._src.interpreters.partial_eval, "get_aval", get_aval2),
+        (jax._src.core, "clean_up_dead_vars", _no_clean_up_dead_vars),
```

Comment:
Good point. In general it's a great idea to document in the code _why_ a function or piece of code is needed (if it isn't immediately obvious).
A good rule of thumb is whenever there is an issue during development, and some new code is added or some.of the initial code is changed, that is probably a good indication that a comment would be helpful.

Reply:
* `get_aval2` corresponds to one of your Jax PR. It is to be able to pass abstract values as arguments
* `_no_clean_up_dead_vars` is another hack required to lower dynamic API programs to MLIR

Reply:
Yes, I mean `_no_clean_up_dead_vars`. Can you elaborate on what the issue is there?

Reply:
Note the comments near these functions.

Reply:
Are we going to contribute a PR that removes `clean_up_dead_vars`, is this meant to be a temporary fix, or should we invest some time to find an alternative that resolves this issue without redefining the function to `return None`?

Reply:
This patch effectively disables the function as a workaround of some bug. I think we need to investigate before sending the PR because I doubt they accept it as-is.